src/minimization.cpp

```diff
@@ -0,0 +1,211 @@
+    }
+  return amountOfLayers;
+}
+
+std::vector<std::vector<TreeNode::nodeId>> Net::getNodesByLayer() {
+  int amountOfLayers = getAmountOfLayers(this->nodes);
+  std::vector<std::vector<TreeNode::nodeId>> nodesByLayer(amountOfLayers + 1);
+  for (TreeNode &node: nodes) {
+    std::vector<TreeNode::nodeId> &layer = nodesByLayer[node.layer];
+    layer.push_back(node.id);
+    node.number = layer.size() - 1;
+  }
+  return nodesByLayer;
+}
+
+int getPortIndex(TreeNode::nodeId id, TreeNode node) {
+  for (int i = 0; i < node.succ.size(); ++i) {
+    if (node.succ[i] == id)
+      return i;
+    }
+  for (int i = 0; i < node.pred.size(); ++i) {
+    if (node.pred[i] == id)
+      return i;
+    }
+}
+
+int crossCounting(Net &net, std::vector<std::vector<TreeNode::nodeId>> &tempNodesByLayer) {
+  int crossNum = 0;
+  for (int i = 1; i < tempNodesByLayer.size(); ++i) {
+    std::vector<std::pair<int, int>> edges;
+    for (int j = 0; j < tempNodesByLayer[i].size(); ++j) {
+      TreeNode node = *net.getNode(tempNodesByLayer[i][j]);
+      for (TreeNode::nodeId &predecessor : node.pred) {
+        TreeNode *predecessorNode = net.getNode(predecessor);
+        if (predecessorNode->layer > node.layer)
+          continue;
+        edges.emplace_back(predecessorNode->number, j);
+      }
+      for (TreeNode::nodeId &successor : node.succ) {
+        TreeNode *successorNode = net.getNode(successor);
+        if (successorNode->layer > node.layer)
+          continue;
+        edges.emplace_back(successorNode->number, j);
+      }
+    }
+    std::sort(edges.begin(), edges.end());
+    for (int j = 0; j < edges.size() - 1; ++j) {
+      for (int k = 0; k < edges.size() - 1; ++k) {
+        if (edges[k].second > edges[k + 1].second) {
+          std::swap(edges[k], edges[k + 1]);
+          crossNum += 1;
+        }
+      }
+    }
+  }
+  return crossNum;
+}
+
+void sortPorts(Net& net, std::vector<std::vector<TreeNode::nodeId>>& nodesByLayer) {
+  for (int i = 1; i < nodesByLayer.size(); ++i) {
+    for (int j = 0; j < nodesByLayer[i].size(); ++j) {
+      TreeNode *node = net.getNode(nodesByLayer[i][j]);
+      std::vector<std::pair<float, TreeNode::nodeId>> bPred, bSucc;
+      for (TreeNode::nodeId &predecessor : node->pred) {
+        TreeNode *predecessorNode = net.getNode(predecessor);
+        float bValue = forwardRankDefinition(*predecessorNode, predecessorNode->number,
+          getPortIndex(node->id, *predecessorNode)) /
+          float(node->succ.size() + node->pred.size());
+        bPred.push_back({ bValue, predecessorNode->id });
+      }
+      for (TreeNode::nodeId &successor : node->succ) {
+        TreeNode *successorNode = net.getNode(successor);
+        float bValue = backwardRankDefinition(*successorNode, successorNode->number,
+          getPortIndex(node->id, *successorNode)) /
+          float(node->succ.size() + node->pred.size());
+        bSucc.push_back({ bValue, successorNode->id });
+      }
```

Comment:
Code duplication. This is actually the case to use a boolean flag: create a procedure with a boolean flag to use either `forwardRankDefinition` or `backwardRankDefinition` internally, then call the procedure twice with proper arguments.